src/crmf/request.rs

```diff
@@ -12,312 +12,109 @@ use num_bigint::{BigInt, BigUint};
 }
 
-/// The `CertRequest` type is defined in [RFC 4211 Section 5].
-///
-/// ```text
-///   CertRequest ::= SEQUENCE {
-///       certReqId     INTEGER,
-///       -- ID for matching request and reply
-///       certTemplate  CertTemplate,
-///       -- Selected fields of cert to be issued
-///       controls      Controls OPTIONAL }
-///       -- Attributes affecting issuance
-/// ```
-///
-/// [RFC 4211 Section 5]: https://www.rfc-editor.org/rfc/rfc4211#section-5
-#[derive(Clone, Debug, Eq, PartialEq, Hash)]
-pub struct CertRequest {
-    pub cert_req_id: BigInt,
-    pub cert_template: CertTemplate,
-    pub controls: Option<Controls>,
-}
-
-impl DerWrite for CertRequest {
-    fn write(&self, writer: DERWriter) {
-        writer.write_sequence(|writer| {
-            self.cert_req_id.write(writer.next());
-            self.cert_template.write(writer.next());
-            if let Some(controls) = self.controls.as_ref() {
-                controls.write(writer.next());
-            }
-        })
-    }
+derive_sequence_of!{
+    /// AttributeSeq corresponds to the type that is inlined in the CertReqMsg
+    /// definition for the regInfo field, as shown below:
+    /// ```text
+    ///       regInfo   SEQUENCE SIZE(1..MAX) OF
+    ///           SingleAttribute{{RegInfoSet}} OPTIONAL }
+    /// ```
+    Attribute<'static> => AttributeSeq
 }
 
-impl BERDecodable for CertRequest {
-    fn decode_ber(reader: BERReader) -> ASN1Result<Self> {
-        reader.read_sequence(|reader| {
-            let cert_req_id = <BigInt as BERDecodable>::decode_ber(reader.next())?;
-            let cert_template = <CertTemplate as BERDecodable>::decode_ber(reader.next())?;
-            let controls = reader.read_optional(|r| Controls::decode_ber(r))?;
-            Ok(CertRequest {
-                cert_req_id,
-                cert_template,
-                controls,
-            })
-        })
+derive_sequence! {
+    /// The `CertTemplate` type is defined in [RFC 4211 Section 5].
+    ///
+    /// ```text
+    ///   CertTemplate ::= SEQUENCE {
+    ///       version      [0] Version               OPTIONAL,
+    ///       serialNumber [1] INTEGER               OPTIONAL,
+    ///       signingAlg   [2] AlgorithmIdentifier{SIGNATURE-ALGORITHM,
+    ///                            {SignatureAlgorithms}}   OPTIONAL,
+    ///       issuer       [3] Name                  OPTIONAL,
+    ///       validity     [4] OptionalValidity      OPTIONAL,
+    ///       subject      [5] Name                  OPTIONAL,
+    ///       publicKey    [6] SubjectPublicKeyInfo  OPTIONAL,
+    ///       issuerUID    [7] UniqueIdentifier      OPTIONAL,
+    ///       subjectUID   [8] UniqueIdentifier      OPTIONAL,
+    ///       extensions   [9] Extensions{{CertExtensions}}  OPTIONAL }
+    /// ```
+    ///
+    /// [RFC 4211 Section 5]: https://www.rfc-editor.org/rfc/rfc4211#section-5
+    ///
+    /// Note: `issuer` and `subject` are EXPLICIT tagged, because they are type of CHOICE
```

Comment:
How does explicit or implicit work? I don't see what in the RFC tells you if a field is explicit or implicit.

Reply:
Updated comments for explaining this:
```rust
    /// Tags are IMPLICIT TAG according to [RFC4211#appendix-B](https://datatracker.ietf.org/doc/html/rfc4211#appendix-B),
    /// but `issuer` and `subject` are EXPLICIT tagged because they are ASN.1 type of `CHOICE`.
```